quota/cache/cache_test.go

```diff
@@ -0,0 +1,205 @@
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx := context.Background()
+	tokens := 5
+	for _, want := range []error{nil, errors.New("llama ate all tokens")} {
+		mock := quota.NewMockManager(ctrl)
+		qm := NewCachedManager(mock, batchSize, maxEntries)
+
+		mock.EXPECT().PutTokens(ctx, tokens, specs).Return(want)
+		if err := qm.PutTokens(ctx, tokens, specs); err != want {
+			t.Errorf("PutTokens() returned err = %v, want = %v", err, want)
+		}
+
+		mock.EXPECT().ResetQuota(ctx, specs).Return(want)
+		if err := qm.ResetQuota(ctx, specs); err != want {
+			t.Errorf("ResetQuota() returned err = %v, want = %v", err, want)
+		}
+	}
+}
+
+func TestCachedManager_GetTokens_CachesTokens(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx := context.Background()
+	tokens := 3
+	mock := quota.NewMockManager(ctrl)
+	mock.EXPECT().GetTokens(ctx, tokens+batchSize, specs).Times(2).Return(nil)
+
+	qm := NewCachedManager(mock, batchSize, maxEntries)
+
+	// Quota requests happen in tokens+batchSize steps, so that batchSize tokens get cached after
+	// the the request is satisfied.
+	// Therefore, the call pattern below is satisfied by just 2 underlying GetTokens() calls.
+	calls := []int{tokens, batchSize, tokens, batchSize / 2, batchSize / 2}
+	for i, call := range calls {
+		if err := qm.GetTokens(ctx, call, specs); err != nil {
+			t.Fatalf("GetTokens() returned err = %v (call #%v)", err, i+1)
+		}
+	}
+}
+
+func TestCachedManager_GetTokens_EvictsCache(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mock := quota.NewMockManager(ctrl)
+	mock.EXPECT().GetTokens(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes().Return(nil)
+
+	ctx := context.Background()
+	qm := NewCachedManager(mock, batchSize, maxEntries)
+
+	originalNow := now
+	defer func() { now = originalNow }()
+	currentTime := time.Time{}
+	now = func() time.Time {
+		currentTime = currentTime.Add(1 * time.Second)
+		return currentTime
+	}
+
+	// Ensure Global quotas are the oldest, we don't want those to get evicted regardless of age.
+	tokens := 5
+	if err := qm.GetTokens(ctx, tokens, []quota.Spec{
+		{Group: quota.Global, Kind: quota.Read},
+		{Group: quota.Global, Kind: quota.Write},
+	}); err != nil {
+		t.Fatalf("GetTokens() returned err = %v", err)
+	}
+
+	// Fill the cache up to maxEntries
+	firstTree := int64(10)
+	tree := firstTree
+	for i := 0; i < maxEntries-2; i++ {
+		if err := qm.GetTokens(ctx, tokens, specForTree(tree)); err != nil {
+			t.Fatalf("GetTokens() returned err = %v (i = %v)", err, i)
+		}
+		tree++
+	}
+
+	// All entries added now must cause eviction of the oldest entries
+	evicts := 5
+	for i := 0; i < evicts; i++ {
+		mock.EXPECT().PutTokens(ctx, batchSize, specForTree(firstTree+int64(i))).Return(nil)
+		if err := qm.GetTokens(ctx, tokens, specForTree(tree)); err != nil {
+			t.Fatalf("GetTokens() returned err = %v (i = %v)", err, i)
+		}
+		tree++
+	}
+
+	qm.(*manager).wait()
```

Comment:
The test will deadlock if an unexpected function on the mock is called by an eviction goroutine (or any of the goroutines that they start), because the goroutine will panic and therefore not call `wg.Done()` (it isn't deferred) . That's not *necessarily* worth defending against, but something to be aware of.